zproject/backends.py

```diff
@@ -2618,6 +2678,10 @@ def auth_complete(self, *args: Any, **kwargs: Any) -> Optional[HttpResponse]:
 
         if isinstance(saml_document, SAMLRequest):
```

Comment:
Should this have a check of the form `saml_document.is_logout_response` too? I feel like the code flow around here is a bit hard to read around what we expect the possibilities to be, and it'd be good to have stronger guarantees via assertions or whatnot that what we're processing is the right type of SAML request.

Reply:
Hmm, the point here is that there's only one  type of `SAMLRequest` that we can take, so we don't need to branch here, and thus also don't need to do any additional checks. We could make an assertion here about some basic parsing of the XML confirming it's a LogoutRequest, but it would be redundant -  i suppose we could add it if we feel it'll improve readability? I tried adding some comments to explain this for now:
```
        if isinstance(saml_document, SAMLRequest):
            # We're a Service Provider, so the only SAMLRequest we accept is a LogoutRequest. Thus
            # we can proceed with process_logout and it's for the lower level libraries to reject it
            # if it's not a valid LogoutRequest.
            return self.process_logout(subdomain, idp_name)
        elif isinstance(saml_document, SAMLResponse) and saml_document.is_logout_response():
            # As a Service Provider, we process SAMLResponse which can be either LogoutResponse
            # or an authentication response.
            return SAMLSPInitiatedLogout.process_logout_response(saml_document, idp_name)
```

Reply:
I think in security code it's almost always good to have defensive assertions around this sort of thing. The fact that we have a check for whether we received a SAMLRequest, and then we interpret it as definitely a logout request, feels like a code smell.

I think it looks like `self._create_saml_auth` is the thing that will actually validate that it is the right request type? I can't really tell from reading the code, and I think we should make it super clear in this comment how that fits together.

Reply:
Maybe to describe more clearly my concern, I feel like we've got 3 cases here, and only in the middle case are we clearly checking that the SAML document that we received is the right type of document.

The part I'm most worried about is actually the block after this, where we have `assert isinstance(saml_document, SAMLResponse)` -- that asserts it's a SAML response, but there's no comment or check on whatever type we're expecting here.

Reply:
I pushed a commit with a more extensive comment there, perhaps that makes things clearer there?

> I think it looks like self._create_saml_auth is the thing that will actually validate that it is the right request type? I can't really tell from reading the code, and I think we should make it super clear in this comment how that fits together.

No, that would indeed make all of this a probable security bug.  All the validation is done in the respective `process_*` functions that we call.  It's not really plausible imo to make security assertions anyway - because it would be hard to really process the XML and determine it's truly what it seems to be - that's essentially full validation.  We aim to not do anything that involves trust until `process_*` gets called. And it's fine to provide the wrong SAML document to the wrong `process_*` function because it'll just fail validation.

> The part I'm most worried about is actually the block after this, where we have assert isinstance(saml_document, SAMLResponse) -- that asserts it's a SAML response, but there's no comment or check on whatever type we're expecting here.

I added a comment at this assert now. Here as well the idea is that this will be validated in python3-saml's `process_response` (https://github.com/SAML-Toolkits/python3-saml/blob/master/src/onelogin/saml2/auth.py#L114) ultimately , which happens in `super().authenticate()` and we're not trusting it until that call. This is just vanilla `SAMLAuthBackend` as originally implemented (in `python-social-auth`) - it just takes the `SAMLResponse` and passes it to `python3-saml` to handle, assuming it's an authentication SAMLResponse. If a confused, or malicious, client passed a different kind of SAMLResponse, that's fine, `python3-saml` will just reject it.

Reply:
We could wrap and rename these methods/functions to be named `validate_and_process_*()` instead if that's going to make the code meaningfully clearer in intent